### ISSUE TITLE
Some computed methods need non-contextual dependencies, it's not DRY to always list them

### ISSUE THREAD

**Issue by sjmiles** [Polymer/polymer#3299]
Consider a translation scheme:

```
{{translate(translator, 'Hello World')}}

translate: function(translator, text, params) {
  return translator(text, params);
}
```

The design is that the 'translator` engine is injected by some means (simple property set, e.g.), therefore we must depend on 'translator' to trigger the relevant binding updates. Having to include this static name in every translate binding is a hassle.

One possibility is to add a way to register static dependencies for a computation.

```
{{translate('Hello World')}}

depends: {
  translate: 'translator'
},
translate: function(text, params) {
  return this.translator(text, params);
}
```

Something like that.

**Comment by robdodson** [comment]
bonus points if this plays nice with behaviors

**Comment by TimvdLippe** [comment]
I submitted an implementation in #3302 .

The implementation is a bit different that the proposal. I discussed this with @sjmiles and he agreed it was "a superior solution".

The example in the docs is:
```
<dom-module id="x-computed-with-dependency">
  <template>
    <span>{{methodName('literal')}}</span>
  </template>
  <script>
   Polymer({
     is: 'element-with-dependency',

     properties: {
       property1: {...},
       property2: {...}
     },

     dependencies: {
       methodName: ['property1', 'property2']
     },

     methodName: function(property1, property2, literal) {}
   });
 </script>
</dom-module>
```

I have not tested behavior support, but I think that could be implemented fairly easily.

**Comment by robdodson** [comment]
Thanks for looking into this @TimvdLippe!

> Could you elaborate on the specifications for behavior support? E.g. What should happen when behavior and child both declare the same or different dependencies?

For my use case I'd want the implementation for `methodName` and `dependencies` to both be part of the behavior that gets mixed in to the element.

As for parent and child declaring same or different dependencies, I haven't thought about that much... Maybe parent just overrides the required dependencies of child?

**Comment by samccone** [comment]
Hey all, so this is the basic patten to have partially applied dependent property methods in polymer today:

![screen shot 2016-01-16 at 1 36 53 pm](https://cloud.githubusercontent.com/assets/883126/12374635/418317b6-bc56-11e5-84aa-6fb2aaaf13d0.png)


In general I dislike 1 major piece of this... the string based observer DSL.
It is a very brittle pattern in a large scale application to depend on strings matching up with props.

---

Now back to this proposal, adding more string based DSL dependency injection esk behavior is pulling us even further from a "vanillajs" world, and potentially increasing the brittleness of components


-- just some thoughts, I do not have a great idea for a clean solution to this.. but I agree it is really hard and not user friendly atm

**Comment by sjmiles** [comment]
@samccone: unless I'm missing something, your pattern can't update the text in the `<p>` when the 'translate' method changes. This is a critical bit provide by the 'dependencies' solution.

**Comment by sjmiles** [comment]
@TimvdLippe: to follow current behaviors rules, the 'dependencies' objects should be (effectively) combined, everything else follows the standard inheritance model (i.e. any method implementations are 'last one wins', where 'last' means left-to-right in behavior arrays and the user's prototype at the end [the very last]).

**Comment by samccone** [comment]
@sjmiles yep your right nm

**Comment by TimvdLippe** [comment]
@sjmiles @robdodson Behavior support should now be added to the dependencies maps. Please check out [this test](https://github.com/Polymer/polymer/pull/3302/files#diff-0464f93d5899ec9b84725e4127e3b403R586)

**Comment by robdodson** [comment]
Cool so it looks like this covers this aspect:
> Conflicts in dependencies should also be handled as 'last one wins'.

Are you also thinking of adding merged depenencies?

**Comment by TimvdLippe** [comment]
@robdodson Merging of dependencies was already possible, but I forgot to add an explicit check for it. I have updated the test to show merging of dependencies works correctly.

**Comment by robdodson** [comment]
cool! I'll defer to @sjmiles for approval

**Comment by sjmiles** [comment]
LGTM, will run it by @sorvell and/or @kevinpschaaf for extra care.

**Comment by kaste** [comment]
I'm a bit confused why this is called `dependencies` when what it does is: Iff this method is used as a view helper function, we _inject_ these properties automatically for you. Or: when used as a view helper, you work with a partial/bound function. Within your normal JS there is nothing special about this method.

**Comment by TimvdLippe** [comment]
@kaste The new behavior of this feature is that annotated computed functions can be updated whenever one of the global arguments is changed. In my eyes, `inject` means the notion of one-time injection of a dependency, whereas this feature guarantees continuous updates. Therefore I would still call it `dependencies` rather than `injectees` or alike. However, if the behavior could be better described we can of course change it.

**Comment by kaste** [comment]
Take #3358 as a rough concept. I really have no idea if it fulfills all and every requirement. On the other side it's in the same ad-hoc style as the rest of the Polymer source code. 

Following example:

```
<dom-module id="x-bind-computed-property">
  <template>
    <div id="check">[[translate('Hello World.')]]</div>
  </template>
  <script>
    Polymer({
      is: 'x-bind-computed-property',
      properties: {
        translate: {
          type: Function,
          computed: '_computeTranslateFn(translator)'
        },
        translator: {
          type: Function,
          value: function() {
            return function(message) {
              return 'translated: ' + message;
            }
          }
        },
      },

      _computeTranslateFn: function(translator) {
        return function(message) {
          return translator(message);
        }
      },

    });
  </script>
</dom-module>
```

As you can see `translate` is a bound function in the sense that it is a property controlled by Polymer. In the example the property is a computed, but this is not required. You could use a complex observer that assigns a new function to `translate`as well. 

As you can see 2: No further configuration is necessary. And although functions which return functions is always a bit meta and an advanced topic, I think any programmer can grab the concept without reading the documentation. I actually tried this before (I wanted to implement the filter function for a dom-repeat in this way), but it wasn't supported (which wasn't documented by the way). So this approach more likely fills a gap in Polymer.

**Comment by kevinpschaaf** [comment]
We had a bunch of discussion on the Polymer team about the pros/cons of the alternate approaches by @TimvdLippe (#3302) and @kaste (#3358).  The upshot is that people actually found it strange after thinking about it that observers/computing functions didn't have _themselves_ as an implicit dependency, which is indeed an alternate way to solve the dependency currying problem for the translate use case as @kaste's PR showed.  So I think I'll re-open #3358 and make a couple more suggestions there for improvements, and see if we can reach agreement on that approach.

**Comment by azakus** [comment]
Fixed with #3358

**Comment by kevinpschaaf** [comment]
For posterity, here's an example using the technique we merged (dynamic computing functions) to implement late-bound string translation via a behavior that results in a pretty pleasant end-user experience: http://jsbin.com/lajuci/edit?html,output
